src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java

```diff
@@ -0,0 +1,522 @@
+            }
+            return new FetchResult<>(shardId, fetchData, failedNodes);
+        }
+    }
+
+    /**
+     * Called by the response handler of the async action to fetch data. Verifies that its still working
+     * on the same cache generation, otherwise the results are discarded. It then goes and fills the relevant data for
+     * the shard (response + failures), issueing a reroute at the end of it to make sure there will be another round
+     * of allocations taking this new data into account.
+     */
+    protected synchronized void processAsyncFetch(ShardId shardId, long cacheGeneration,
+                                                  T[] responses, FailedNodeException[] failures) {
+        if (this.cacheGeneration.get() != cacheGeneration) {
+            // the caches got replaced on us since master changed, this fetch is not relevant
+            return;
+        }
+        Map<String, NodeEntry<T>> shardCache = cache.get(shardId);
+        if (shardCache == null) {
+            // node longer with us, nothing to do
+            return;
+        }
+        if (responses != null) {
+            for (T response : responses) {
+                NodeEntry<T> nodeEntry = shardCache.get(response.getNode().getId());
+                // if the entry is there, and not marked as failed already, process it
+                if (nodeEntry != null && nodeEntry.isFailed() == false) {
+                    nodeEntry.doneFetching(response);
+                }
+            }
+        }
+        if (failures != null) {
+            for (FailedNodeException failure : failures) {
+                NodeEntry<T> nodeEntry = shardCache.get(failure.nodeId());
+                // if the entry is there, and not marked as failed already, process it
+                if (nodeEntry != null && nodeEntry.isFailed() == false) {
+                    Throwable unwrappedCause = ExceptionsHelper.unwrapCause(failure.getCause());
+                    // if the request got rejected or timed out, we need to try it again next time...
+                    if (unwrappedCause instanceof EsRejectedExecutionException || unwrappedCause instanceof ReceiveTimeoutTransportException || unwrappedCause instanceof ElasticsearchTimeoutException) {
+                        nodeEntry.restartFetching();
+                    } else {
+                        logger.warn("{}: failed to list shard for {} on node [{}]", failure, shardId, getClass().getSimpleName(), failure.nodeId());
+                        nodeEntry.doneFetching(failure.getCause());
+                    }
+                }
+            }
+        }
+        reroute(shardId, "post_response");
+    }
+
+    /**
+     * Schedules a reroute on the cluster update thread.
+     */
+    protected void reroute(ShardId shardId, String reason) {
+        clusterService.submitStateUpdateTask("done_fetch_shard_data for shard " + shardId + ", reasons (" + reason + ")", Priority.HIGH, new ClusterStateUpdateTask() {
```

Comment:
can we factor this out of this class and remove the useless `@Inject` from the `AsyncShardFetch.Started` &  `AsyncShardFetch.Store`, make the `GatewayAllocator` instantiate these classes directly. We can then have a simple `OnShardFetched` callback that we can pass as a ctor argument that does the reroute for us. Similarly, the `GatewayAllocator` should also register the cluster change listener and if you ask me I'd drop all this generation thing  and replace the entire cache if we are getting a new master in the `GatewayAllocator`.